go/registry/api/sanity_check.go

```diff
@@ -147,43 +175,151 @@ func SanityCheckNodes(
 			nodeLookup,
 		)
 		if err != nil {
-			return fmt.Errorf("registry: sanity check failed for node: %s, error: %w", n.ID.String(), err)
+			return nil, fmt.Errorf("registry: node sanity check failed: ID: %s, error: %w", n.ID.String(), err)
 		}
 
 		// Add validated node to nodeLookup.
 		nodeLookup.nodes[node.Consensus.ID] = node
 		nodeLookup.nodes[node.P2P.ID] = node
+		nodeLookup.nodesList = append(nodeLookup.nodesList, node)
 
 		var h = hash.Hash{}
 		h.FromBytes(node.Committee.Certificate)
 		nodeLookup.nodesCertHashes[h] = node
 	}
 
+	return nodeLookup, nil
+}
+
+// SanityCheckStake ensures entities' stake accumulator claims are consistent
+// with general state and entities have enough stake for themselves and all
+// their registered nodes and runtimes.
+func SanityCheckStake(
+	entities []*entity.Entity,
+	accounts map[signature.PublicKey]*staking.Account,
+	nodes []*node.Node,
+	runtimes []*Runtime,
+	stakeThresholds map[staking.ThresholdKind]quantity.Quantity,
+) error {
+	// Entities' escrow accounts for checking claims and stake.
+	generatedEscrows := make(map[signature.PublicKey]*staking.EscrowAccount)
+
+	// Generate escrow account for all entities.
+	for _, entity := range entities {
+		acct, ok := accounts[entity.ID]
+		if !ok {
+			return fmt.Errorf("registry: sanity check failed: no account associated with entity with ID: %v", entity.ID)
+		}
+
+		// Generate an escrow account with the same active balance and shares number.
+		escrow := &staking.EscrowAccount{
+			Active: staking.SharePool{
+				Balance:     acct.Escrow.Active.Balance,
+				TotalShares: acct.Escrow.Active.TotalShares,
+			},
+		}
+
+		// Add entity stake claim.
+		escrow.StakeAccumulator.AddClaimUnchecked(StakeClaimRegisterEntity, []staking.ThresholdKind{staking.KindEntity})
+
+		generatedEscrows[entity.ID] = escrow
+	}
+
+	for _, node := range nodes {
+		// Add node stake claims.
+		generatedEscrows[node.EntityID].StakeAccumulator.AddClaimUnchecked(StakeClaimForNode(node.ID), StakeThresholdsForNode(node))
```

Comment:
it's already checked that `node.EntityID` is a valid entity ID, right?

Reply:
Yes, it is already checked by `SanityCheckEntities()` which in turn calls `VerifyRegisterEntityArgs()` for each signed entity which finally checks the ID in:
https://github.com/oasislabs/oasis-core/blob/b9fe8b7e4ea570b9aa69ec549ec1f3da5ffdcd70/go/registry/api/api.go#L363-L370